File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/modulejournal/login/ModuleJournalLoginFragment.java
package com.vereshchagin.nikolay.stankinschedule.modulejournal.login;

import android.animation.ValueAnimator;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.NavOptions;
import androidx.navigation.Navigation;

import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.vereshchagin.nikolay.stankinschedule.R;
import com.vereshchagin.nikolay.stankinschedule.modulejournal.network.ModuleJournalError;
import com.vereshchagin.nikolay.stankinschedule.modulejournal.network.ModuleJournalService;
import com.vereshchagin.nikolay.stankinschedule.utils.CommonUtils;
import com.vereshchagin.nikolay.stankinschedule.utils.TextWatcherWrapper;

import java.lang.ref.WeakReference;

/**
 * Фрагмент авторизации в модульный журнал.
 */
public class ModuleJournalLoginFragment extends Fragment
        implements View.OnClickListener {

    private TextInputLayout mLoginFieldLayout;
    private TextInputEditText mLoginFieldEditText;

    private TextInputLayout mPasswordFieldLayout;
    private TextInputEditText mPasswordFieldEditText;

    private Button mSignInButton;
    private Button mForgotPasswordButton;

    private ProgressBar mLoginLoading;
    private ValueAnimator mLoginLoadingAnimator;
    private int mLoginLoadingHeight;

    /**
     * ViewModel для авторизации.
     */
    private ModuleJournalLoginModel mModuleJournalLoginModel;

    public ModuleJournalLoginFragment() {
        super();

        mLoginLoadingAnimator = new ValueAnimator();
        mLoginLoadingAnimator.setDuration(300);
        mLoginLoadingAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                mLoginLoading.getLayoutParams().height = (int) animation.getAnimatedValue();
                mLoginLoading.requestLayout();
            }
        });
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_module_journal_login, container, false);

        mLoginFieldLayout = view.findViewById(R.id.mj_login_layout);
        mLoginFieldEditText = view.findViewById(R.id.mj_login);

        mPasswordFieldLayout = view.findViewById(R.id.mj_password_layout);
        mPasswordFieldEditText = view.findViewById(R.id.mj_password);

        DataTextWatcher loginWatcher = new DataTextWatcher(mLoginFieldLayout);
        mLoginFieldEditText.addTextChangedListener(loginWatcher);

        DataTextWatcher passwordWatcher = new DataTextWatcher(mPasswordFieldLayout);
        mPasswordFieldEditText.addTextChangedListener(passwordWatcher);

        mSignInButton = view.findViewById(R.id.mj_sign_in);
        mSignInButton.setOnClickListener(this);

        mForgotPasswordButton = view.findViewById(R.id.mj_forgot_password);
        mForgotPasswordButton.setOnClickListener(this);

        mLoginLoading = view.findViewById(R.id.mj_login_loading);
        mLoginLoadingHeight = getResources().getDimensionPixelOffset(R.dimen.horizontal_loading_height);

        mModuleJournalLoginModel = new ViewModelProvider(this,
                new ModuleJournalLoginModel.Factory(getActivity().getApplication()))
                .get(ModuleJournalLoginModel.class);

        mModuleJournalLoginModel.stateData().observe(getViewLifecycleOwner(), new Observer<ModuleJournalLoginModel.State>() {
            @Override
            public void onChanged(ModuleJournalLoginModel.State state) {
                stateChanged(state);
            }
        });

        return view;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        // скрыть клавиатуру перед выходом
        if (getActivity() != null) {
            InputMethodManager manager = ContextCompat.getSystemService(getActivity(),
                    InputMethodManager.class);

            View currentFocusedView = getActivity().getCurrentFocus();
            if (currentFocusedView != null && manager != null) {
                manager.hideSoftInputFromWindow(currentFocusedView.getWindowToken(),
                        InputMethodManager.HIDE_NOT_ALWAYS);
            }
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            // авторизация
            case R.id.mj_sign_in: {
                if (!fieldsIsCorrect()) {
                    return;
                }

                setLoading(true);

                mLoginFieldEditText.clearFocus();
                mPasswordFieldEditText.clearFocus();

                Editable login = mLoginFieldEditText.getText();
                Editable password = mPasswordFieldEditText.getText();

                mModuleJournalLoginModel.singIn(login == null ? "" : login.toString(),
                        password == null ? "" : password.toString());

                break;
            }
            // забыт пароль
            case R.id.mj_forgot_password: {
                Context context = getContext();
                if (context != null) {
                    CommonUtils.openBrowser(context, ModuleJournalService.URL);
                }
                break;
            }
        }
    }

    /**
     * Спускает навигацию до фрагмента с модульным журналом.
     */
    private void popToModuleJournal() {
        if (getActivity() != null) {
            NavController controller = Navigation.findNavController(getActivity(), R.id.nav_host);

            NavOptions options = new NavOptions.Builder()
                    .setPopUpTo(R.id.nav_module_journal_login_fragment, true)
                    .setEnterAnim(R.anim.nav_default_enter_anim)
                    .setExitAnim(R.anim.nav_default_exit_anim)
                    .setPopEnterAnim(R.anim.nav_default_pop_enter_anim)
                    .setPopExitAnim(R.anim.nav_default_pop_exit_anim)
                    .build();

            controller.navigate(R.id.nav_module_journal_fragment, null, options);
        }
    }

    /**
     * Устанавливает загрузку в фрагмент, пока происходит авторизация.
     * @param isLoading true - включить, false - выключить.
     */
    private void setLoading(boolean isLoading) {
        mLoginFieldEditText.setEnabled(!isLoading);
        mPasswordFieldEditText.setEnabled(!isLoading);

        mSignInButton.setEnabled(!isLoading);
        mForgotPasswordButton.setEnabled(!isLoading);

        int target = isLoading ? mLoginLoadingHeight : 0;
        mLoginLoadingAnimator.setIntValues(mLoginLoading.getMeasuredHeight(), target);
        mLoginLoadingAnimator.start();
    }

    /**
     * Проверяет поля для ввода данных на корректность.
     * @return true - поля заполнены правильно, иначе false.
     */
    private boolean fieldsIsCorrect() {
        Editable login = mLoginFieldEditText.getText();
        Editable password = mPasswordFieldEditText.getText();

        String loginText = login != null ? login.toString() : null;
        String passwordText = password != null ? password.toString() : null;

        if (loginText == null || passwordText == null) {
            return false;
        }

        if (loginText.isEmpty()) {
            mLoginFieldLayout.setErrorEnabled(true);
            mLoginFieldLayout.setError(getString(R.string.mj_empty_filed));
            return false;
        }

        if (passwordText.isEmpty()) {
            mPasswordFieldLayout.setEnabled(true);
            mPasswordFieldLayout.setError(getString(R.string.mj_empty_filed));
            return false;
        }

        return true;
    }

    /**
     * Вызывается, когда поменялось состояние авторизации.
     * @param state состояние.
     */
    private void stateChanged(@NonNull ModuleJournalLoginModel.State state) {
        switch (state) {
            case AUTHORIZED: {
                popToModuleJournal();
                break;
            }
            case LOADING: {
                setLoading(true);
                break;
            }
            case ERROR: {
                setLoading(false);

                ModuleJournalError error = mModuleJournalLoginModel.error();

                if (error == null) {
                    return;
                }

                String title = error.errorTitle();
                if (title == null) {
                    title = getString(error.errorTitleRes());
                }

                String description = error.errorDescription();
                if (description == null) {
                    description = getString(error.errorDescriptionRes());
                }

                new AlertDialog.Builder(getContext())
                        .setTitle(title)
                        .setMessage(description)
                        .setNeutralButton(R.string.ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                            }
                        })
                        .create()
                        .show();
                break;
            }
            case WAIT: {
                setLoading(false);
                break;
            }
        }
    }


    /**
     * Watcher для проверки полей ввода данных.
     */
    private class DataTextWatcher extends TextWatcherWrapper {

        private WeakReference<TextInputLayout> mFieldLayout;

        DataTextWatcher(@NonNull TextInputLayout fieldLayout) {
            mFieldLayout = new WeakReference<>(fieldLayout);
        }

        @Override
        public void onTextChanged(@NonNull String s) {
            TextInputLayout fieldLayout = mFieldLayout.get();
            if (fieldLayout  == null) {
                return;
            }

            if (s.length() == 0) {
                fieldLayout.setErrorEnabled(true);
                fieldLayout.setError(getString(R.string.mj_empty_filed));
            } else {
                fieldLayout.setErrorEnabled(false);
                fieldLayout .setError(null);
            }
        }
    }
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/ui/modulejournal/login/ModuleJournalLoginFragment.kt
package com.vereshchagin.nikolay.stankinschedule.ui.modulejournal.login

import android.animation.ValueAnimator
import android.os.Bundle
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.core.widget.doOnTextChanged
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavOptions
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.vereshchagin.nikolay.stankinschedule.R
import com.vereshchagin.nikolay.stankinschedule.databinding.FragmentModuleJournalLoginBinding
import com.vereshchagin.nikolay.stankinschedule.ui.BaseFragment
import com.vereshchagin.nikolay.stankinschedule.utils.CommonUtils
import com.vereshchagin.nikolay.stankinschedule.utils.Constants
import com.vereshchagin.nikolay.stankinschedule.utils.ExceptionUtils
import com.vereshchagin.nikolay.stankinschedule.utils.State
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest

/**
 * Фрагмент входа в модульный журнал.
 */
@AndroidEntryPoint
class ModuleJournalLoginFragment :
    BaseFragment<FragmentModuleJournalLoginBinding>(FragmentModuleJournalLoginBinding::inflate),
    View.OnClickListener {

    /**
     * ViewModel фрагмента.
     */
    private val viewModel: ModuleJournalLoginViewModel by viewModels()

    /**
     * Аниматор прогресса авторизации в модульном журнале.
     */
    private val loadAnimator = ValueAnimator().apply {
        duration = 300
        addUpdateListener(this@ModuleJournalLoginFragment::loadAnimatorUpdate)
    }

    private var loadingHeight = 50

    override fun onPostCreateView(savedInstanceState: Bundle?) {

        loadingHeight = resources.getDimensionPixelOffset(R.dimen.horizontal_loading_height)

        binding.mjLogin.doOnTextChanged { text: CharSequence?, _: Int, _: Int, _: Int ->
            if (text != null) {
                val isError = text.isEmpty()
                binding.mjLoginLayout.error = if (isError) {
                    getString(R.string.mj_empty_filed)
                } else {
                    null
                }
            }
        }

        binding.mjPassword.doOnTextChanged { text: CharSequence?, _: Int, _: Int, _: Int ->
            if (text != null) {
                val isError = text.isEmpty()
                binding.mjPasswordLayout.error = if (isError) {
                    getString(R.string.mj_empty_filed)
                } else {
                    null
                }
            }
        }

        binding.mjPassword.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                signIn()
                true
            } else {
                false
            }
        }

        binding.mjSignIn.setOnClickListener(this)
        binding.mjForgotPassword.setOnClickListener(this)

        // статус авторизации
        lifecycleScope.launchWhenCreated {
            viewModel.authorizedState.collectLatest { state ->
                when (state) {
                    is State.Loading -> {
                        setLoadState(true)
                    }
                    is State.Success -> {
                        // успешно авторизован
                        if (state.data) {
                            loadAnimator.end()
                            navigateToModuleJournal()
                        }
                        setLoadState(false)
                    }
                    is State.Failed -> {
                        val description =
                            ExceptionUtils.errorDescription(state.error, requireContext())

                        MaterialAlertDialogBuilder(requireContext())
                            .setTitle(R.string.error)
                            .setMessage(description)
                            .setNeutralButton(R.string.ok) { dialog, _ ->
                                dialog.dismiss()
                            }
                            .show()

                        setLoadState(false)
                    }
                }
            }
        }
    }

    override fun onDestroyView() {
        loadAnimator.end()
        super.onDestroyView()
        // скрыть клавиатуру перед выходом
        hideKeyboard()
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            // авторизация
            R.id.mj_sign_in -> {
                signIn()
            }
            // забыт пароль
            R.id.mj_forgot_password -> {
                CommonUtils.openBrowser(requireContext(), Constants.MODULE_JOURNAL_URL)
            }
        }
    }

    /**
     * Выполнение входа в модульный журнал.
     */
    private fun signIn() {
        if (!fieldsIsCorrect()) {
            return
        }

        setLoadState(true)
        binding.mjLogin.clearFocus()
        binding.mjPassword.clearFocus()

        val login = binding.mjLogin.text.toString()
        val password = binding.mjPassword.text.toString()

        viewModel.signIn(login, password)
    }

    private fun fieldsIsCorrect(): Boolean {
        val login = binding.mjLogin.text.toString()
        val password = binding.mjPassword.text.toString()

        if (login.isEmpty()) {
            binding.mjLoginLayout.isErrorEnabled = true
            binding.mjLoginLayout.error = getString(R.string.mj_empty_filed)
            return false
        }

        if (password.isEmpty()) {
            binding.mjPasswordLayout.isErrorEnabled = true
            binding.mjPasswordLayout.error = getString(R.string.mj_empty_filed)
            return false
        }

        return true
    }

    /**
     * Устанавливает состоянии загрузки в фрагмент.
     */
    private fun setLoadState(load: Boolean) {
        binding.mjLogin.isEnabled = !load
        binding.mjPassword.isEnabled = !load

        binding.mjSignIn.isEnabled = !load
        binding.mjForgotPassword.isEnabled = !load

        startLoadAnimator(load)
    }

    /**
     * Начинает анимацию загрузки при входе в модульный журнал.
     */
    private fun startLoadAnimator(load: Boolean) {
        loadAnimator.apply {
            setIntValues(binding.mjLoginLoading.measuredHeight, if (load) loadingHeight else 0)
            start()
        }
    }

    /**
     * Вызывается, при обновлении значения аниматора загрузки.
     */
    private fun loadAnimatorUpdate(animator: ValueAnimator) {
        // возможен случай, что анимация может завершиться после уничтожения binding
        rawBinding?.let {
            it.mjLoginLoading.layoutParams.height = animator.animatedValue as Int
            it.mjLoginLoading.requestLayout()
        }
    }

    /**
     * Переходит к фрагменту модульного журнала.
     */
    private fun navigateToModuleJournal() {
        val options = NavOptions.Builder()
            .setPopUpTo(R.id.nav_module_journal_login_fragment, true)
            .setEnterAnim(R.anim.nav_default_enter_anim)
            .setExitAnim(R.anim.nav_default_exit_anim)
            .setPopEnterAnim(R.anim.nav_default_pop_enter_anim)
            .setPopExitAnim(R.anim.nav_default_pop_exit_anim)
            .build()

        navigateTo(R.id.nav_module_journal_fragment, options = options)
    }
}